Update the README file (#1)
sergepilipchuk/wpf-isupportparentviewmodel: README.md

# WPF MVVM Framework - Utilize the ISupportParentViewModel Interface

> This example uses the [`ISupportParentViewModel`](https://docs.devexpress.com/WPF/17449/mvvm-framework/viewmodels/viewmodel-relationships-isupportparentviewmodel) interface that exposes a parent view model to its child view models. The interface is implemented automatically in all [`ViewModelBase`](https://docs.devexpress.com/WPF/17351/mvvm-framework/viewmodels/viewmodelbase) descendants.  

Reviewer: ```suggestion
This example uses the [`ISupportParentViewModel`](https://docs.devexpress.com/WPF/17449/mvvm-framework/viewmodels/viewmodel-relationships-isupportparentviewmodel) interface that exposes a parent view model to its child view models. This interface is implemented automatically in all [`ViewModelBase`](https://docs.devexpress.com/WPF/17351/mvvm-framework/viewmodels/viewmodelbase) descendants.  
```

## Implementation Details

> Use the `ViewModelExtensions.ParentViewModel` attached property to associate a parent view model with a child view model. This property is typically applied in XAML to enable the MVVM framework to propagate services and context from a parent view model to its child:

Reviewer: > This property is typically **applied in** XAML to enable the MVVM framework **to propagate services and context** from a parent view model to its child:

I do not understand these two things:

1. What do you mean by "this property is applied in XAML"? Defined in XAML? Something else?
2. What does this mean? "to enable the MVVM framework to propagate services"? First of all, perhaps, there are some redundant words... Second... what do you mean by "propagate" - copy? proliferate? If yes - how? I do not understand.

We need to clarify this paragraph. Definitely.

Author: Reworked version:
Use the `ViewModelExtensions.ParentViewModel` attached property to set a parent view model for a child view model. Set this attached property in XAML on the child view to allow the child view model to access services and data from the parent view model:

Reviewer: OK. One more question here:

> Set this attached property in XAML **on** the child view

What does it mean set the property **on** the child view? Do you mean set this property at the child view level? To whom does this attach property belongs?

Also, I'd like to ask if "child view" and "child view model" are synonyms in this particular case. Are they the same or different?

> **Set** this attached property

Still unclear. What should I assign to this attached property in XAML? Besides, we see a pattern - set ... set.

I would have went into generalities at the very beginning. For example:

> Use the `ViewModelExtensions.ParentViewModel` property to associate parent and child view models.

And then I would have descended into particularities... The second sentence should clearer than it is right now. To my regret, I cannot offer a change because I don't have comprehensive knowledge on this matter.

Bottom line. I'm waiting for your new version, please.

Author: Use the `ViewModelExtensions.ParentViewModel` attached property to associate a child view model with its parent view model. In XAML, specify the `ParentViewModel` attached property for the child view to allow the child view model to access services and data from the parent view model:

Author: This variant should be clear for understanding.

Reviewer: Indeed this version is better. I'll run one more iteration on this, though...

```suggestion
Use the `ViewModelExtensions.ParentViewModel` attached property to associate a child view model with a parent view model. In XAML, specify the `ParentViewModel` property for the child view as follows so it can access services and data from the required parent view model:
```

Again, this is my version based on the information I gathered. You may (and probably should) come up with a better version. Still, my version would work.

> Use the `ViewModelExtensions.ParentViewModel` attached property to associate a child view model with a parent view model. In XAML, specify the `ParentViewModel` property for the child view as follows so it can access services and data from the required parent view model:

Reviewer 2: ```suggestion
Use the `ViewModelExtensions.ParentViewModel` attached property to associate a child view model with a parent view model. In XAML, specify the `ParentViewModel` property for the child view as follows so it can access services and data from the parent view model:
```

```xaml
<local:ChildView 
    dxmvvm:ViewModelExtensions.ParentViewModel="{Binding DataContext, ElementName=LayoutRoot}"/>
```

> > Note: The `ViewModelExtensions.ParentViewModel` attached property is set **after** the child view is initialized. It is not available in the view model constructor.

Reviewer: OK

> > Note: The `ViewModelExtensions.ParentViewModel` attached property is set **after** the child view is initialized. **It** is not available in the view model constructor.

What's "it" is referring to? The child view? The attached property? One more clarity issue is on our hands.

Author: Note: The `ViewModelExtensions.ParentViewModel` attached property is set after the child view is initialized.
The parent view model value is not available in the child view model's constructor.

Reviewer: This note is referring to the example we demonstrate in this repository, true. Therefore, we are referring to **the** [current] child view (the one we demonstrate). As for the reason, we speak in generalities - why do we have to do things this way? Because this property is not available in **any** child view constructor (not only this time).

Think of possible class descendants, for example. They all will behave identically, therefore, the approach we demonstrate is always correct.

Yes, we could have went all for definite. However, our editorial policy dictates our choice -- if we can interpret a case as indefinite, we definitely should.

```suggestion
> Note: The `ViewModelExtensions.ParentViewModel` attached property is set **after** the child view is initialized because the property is not available in a child view constructor.
```

The [`ISupportParentViewModel`](https://docs.devexpress.com/WPF/17449/mvvm-framework/viewmodels/viewmodel-relationships-isupportparentviewmodel) interface allows a child view model to access MVVM services defined at the parent level:

```cs
IMessageBoxService MessageBoxService {
    get { return GetService<IMessageBoxService>(ServiceSearchMode.PreferParents); }
}
```

> In `ViewModelBase` descendants, override the `OnParentViewModelChanged` method to raise property change notifications for `ParentViewModel`:

Reviewer: Wait. You do not raise notifications. You can raise events (or even exceptions; more throw exceptions, but you can also raise them but notifications?) You send them. You send them to whom?

Do you send them to ParentViewModel? One more clarity issue on our hands.

Author: Override the `OnParentViewModelChanged` method to trigger a property change notification for the `ParentViewModel` property and ensure that data bindings update accordingly.

Reviewer: OK. Do I get this right that we send a notification to the parent view model assigned to the `ParentViewModel` property?

Either way, what's the point behind this advice? When/why should I override this method in my descendant? From what I gather, I should do so only when I change the base (underlying) logic of data management and view interaction. That is, when the base `OnParentViewModelChanged` method implementation becomes insufficient (for example, omits certain changes that happen in your own modified model).

Am I right? Thing is, that I need to know what should I do with this information.

Author: >OK. Do I get this right that we send a notification to the parent view model assigned to the ParentViewModel property?

We send a notification ( via INotifyPropertyChanged ) to the ParentViewModel property. The WPF binding engine  mechanism catches this notification to update all bindings that are linked to this property. So, we notify WPF, not the parent ViewModel.

>Either way, what's the point behind this advice? When/why should I override this method in my descendant? From what I gather, I should do so only when I change the base (underlying) logic of data management and view interaction. That is, when the base OnParentViewModelChanged method implementation becomes insufficient (for example, omits certain changes that happen in your own modified model).

Override the `OnParentViewModelChanged` method only if your child view model needs to perform custom actions when its parent view model is assigned or changed.

Reviewer: This is exactly what I wanted to hear. As a result, my take on this paragraph looks as follows:

```suggestion
Override the `OnParentViewModelChanged` method in a `ViewModelBase` descendant only if your child view model implementation includes custom parent view model changes.
```

Of course, this is merely an example. This is my version based on what I've heard on the subject. You can come up with something more suitable than that.
